docs/enhancements/2025-07-18-generic-controller.md

```diff
@@ -0,0 +1,576 @@
+# InGate as a generic Gateway API Controller
+
+InGate was concieved originally to be a simple Gateway API controller,
+allowing users to have a simple and yet comformant implementation of the API
+and also to still make use of existing Ingress APIs.
+
+During the implementation of InGate, and also checking other implementations,
+it was realized that there's no easy way to reproduce existing code into
+our own implementation.
+
+But looking at the existing L7 routing implementations (eg.: kgateway, envoygateway,
+and others) we can conclude that what differ one implementation from the other
+is how the backend is implemented.
+
+For every Gateway API implementation, we see a standard: a comformant way to
+reconcile GatewayClass, then to reconcile Gateway resources, and the *Route
+objects depending on what kind of backend is supported.
+
+When evolving the implementation with new experimental features, like ListenerSet
+there is a need of every implementation to do their own code changes, just to
+be sure that whatever is desired by users on the API is reflected as a backend
+configuration.
+
+This proposal intends to define a well known controller, that can be used by
+backend implementations to do their own configuration, while avoiding to
+re-implement a Gateway API controller every time a new backend wants to support
+it or a new feature is added to Gateway API resources.
+
+It is worth noticing that this approach is not new on Kubernetes world, where
+a CSI, for instance defines how a volume should be created but it is up to each
+backend to program it, or Cluster API, that defines how a Kubernetes cluster should
+be created, and then each Infrastructure Provider can define their own way of
+provisioning and managing the infrastructure for that Kubernetes cluster.
```

Comment:
Do CSI and Cluster API offer shared library code like this though, or just the spec? If not, is there any prior discussion of this concept within those projects?

Reply:
CAPI has its own set of utils: https://github.com/kubernetes-sigs/cluster-api/tree/main/util

And also a well defined definition of contracts (https://cluster-api.sigs.k8s.io/developer/providers/contracts/overview) which is slightly different here, because in the end it is "CAPI" creating other resources that will be reconciled by other controllers.

CSI does something similar: https://kubernetes-csi.github.io/docs/developing.html

In our case, we don't expect that a resource is converted to intermediate resources (eg.: the aggregation of Gateway Listeners + ListenerSets becomes something like MyProviderListeners) because it doesn't make much sense